base/acme/src/main/java/org/dogtagpki/acme/server/ACMEChallengeService.java

```diff
@@ -73,71 +70,36 @@ public Response handlePOST(@PathParam("id") String challengeID, JWS jws) throws
         }
 
-        String type = challenge.getType();
-        logger.info("Challenge Type: " + type);
-
-        ACMEValidator validator = engine.getValidator(type);
-        if (validator == null) {
-            // TODO: generate proper exception
-            throw new Exception("Unsupported challenge type: " + type);
-        }
-
         String challengeStatus = challenge.getStatus();
         if (challengeStatus.equals("pending")) {
-            challenge.setStatus("processing");
-            engine.updateAuthorization(account, authorization);
 
-        } else if (challengeStatus.equals("processing")) {
-            // retrying the challenge, ignore
-
-        } else {
-            // TODO: generate proper exception
-            throw new Exception("Challenge is already " + challengeStatus);
-        }
+            String type = challenge.getType();
+            logger.info("Challenge Type: " + type);
 
-        try {
-            validator.validateChallenge(authorization, challenge);
+            ACMEValidator validator = engine.getValidator(type);
+            if (validator == null) {
+                // TODO: generate proper exception
+                throw new Exception("Unsupported challenge type: " + type);
+            }
 
-        } catch (Exception e) {
-            logger.info("Challenge " + challengeID + " is invalid");
-            challenge.setStatus("invalid");
+            challenge.setStatus("processing");
             engine.updateAuthorization(account, authorization);
-            throw e;
-        }
-
-        logger.info("Challenge " + challengeID + " is valid");
-        challenge.setStatus("valid");
-        challenge.setValidationTime(new Date());
-
-        logger.info("Authorization " + authzID + " is valid");
-        authorization.setStatus("valid");
-
-        engine.updateAuthorization(account, authorization);
 
-        logger.info("Checking associated pending orders for validity");
-        Collection<ACMEOrder> orders =
-            engine.getOrdersByAuthorizationAndStatus(account, authzID, "pending");
+            ACMEChallengeProcessor processor = new ACMEChallengeProcessor(
+                    account,
+                    authorization,
+                    challenge,
+                    validator);
 
-        for (ACMEOrder order : orders) {
-            boolean allAuthorizationsValid = true;
+            // TODO: use thread pool
+            new Thread(processor).start();
 
-            for (String orderAuthzID : order.getAuthzIDs()) {
-                ACMEAuthorization authz = engine.getAuthorization(account, orderAuthzID);
-                if (authz.getStatus().equals("valid")) {
-                    continue;
-                } else {
-                    allAuthorizationsValid = false;
-                    break;
-                }
-            }
+        } else if (challengeStatus.equals("processing")) {
+            // TODO: retry the challenge
```

Comment:
When we implement this, it would need to wake up the thread somehow.  RFC 8555 also says that we SHOULD rate-limit retry requests.

Nothing to be done in this PR, just a related talking-point :)

Reply:
Agreed, but it probably will require further design discussion.